File: polymorphism.js
// ===============================
// 🔹 1. Base Class (Parent)
// ===============================
class Animal {
  constructor(name) {
    this.name = name;
  }

  // Common method (will be overridden)
  speak() {
    console.log(this.name + " makes a sound");
  }
}

// ===============================
// 🔹 2. Child Classes (Overriding)
// ===============================

class Dog extends Animal {
  // Runtime Polymorphism (Method Overriding)
  speak() {
    console.log(this.name + " barks 🐶");
  }
}

class Cat extends Animal {
  speak() {
    console.log(this.name + " meows 🐱");
  }
}

// ===============================
// 🔹 3. Duck Typing (No inheritance)
// ===============================

let robot = {
  name: "Robo",
  speak: function () {
    console.log(this.name + " makes robotic sound 🤖");
  },
};

// ===============================
// 🔹 4. Method Overloading (Simulated)
// ===============================

function greet(name, age) {
  // JS does not support true overloading,
  // so we handle different cases manually
  if (age === undefined) {
    console.log("Hello " + name);
  } else {
    console.log("Hello " + name + ", age: " + age);
  }
}

// ===============================
// 🔥 5. Polymorphism in Action
// ===============================

// Different objects
let dog = new Dog("Tommy");
let cat = new Cat("Kitty");

// Same method call → different behavior
dog.speak(); // Tommy barks 🐶
cat.speak(); // Kitty meows 🐱

// Duck typing → works even without class
robot.speak(); // Robo makes robotic sound 🤖

// ===============================
// 🔹 6. Generic Function (Polymorphic)
// ===============================

function makeSound(entity) {
  // No type check → only behavior matters
  entity.speak();
}

// All objects behave differently with same function
makeSound(dog);
makeSound(cat);
makeSound(robot);

// ===============================
// 🔹 7. Overloading Demo
// ===============================

greet("Mihir"); // Hello Mihir
greet("Mihir", 18); // Hello Mihir, age: 18

File: java/polymorphism.java
//  POLYMORPHISM FULL DEMO IN JAVA

// ===============================
// 🔹 1. Base Class
// ===============================
class Animal {

    // Method to override
    void speak() {
        System.out.println("Animal makes sound");
    }
}

// ===============================
// 🔹 2. Child Class 1 (Overriding)
// ===============================
class Dog extends Animal {

    @Override
    void speak() {
        System.out.println("Dog barks 🐶");
    }
}

// ===============================
// 🔹 3. Child Class 2 (Overriding)
// ===============================
class Cat extends Animal {

    @Override
    void speak() {
        System.out.println("Cat meows 🐱");
    }
}

// ===============================
// 🔹 4. Compile-time Polymorphism
// ===============================
class MathUtils {

    // Method Overloading
    int add(int a, int b) {
        return a + b;
    }

    int add(int a, int b, int c) {
        return a + b + c;
    }
}

// ===============================
// 🔹 5. Main Class
// ===============================
public class Main {
    public static void main(String[] args) {

        // 🔸 Runtime Polymorphism
        Animal a;

        a = new Dog();
        a.speak();  // Dog version

        a = new Cat();
        a.speak();  // Cat version

        // 🔸 Compile-time Polymorphism
        MathUtils m = new MathUtils();

        System.out.println(m.add(5, 10));       // 2 args
        System.out.println(m.add(5, 10, 15));   // 3 args
    }
}